[RPC] Fix flaky test by waiting for async rref calls

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#39012 [RPC] Fix flaky test by waiting for async rref calls**
* #39011 [TensorPipe] Detect duplicate worker names
* #39010 [TensorPipe] Improve serialization
* #38934 [TensorPipe] Close and join TP context at shutdown
* #38933 [TensorPipe] Implement join correctly
* #38930 [TensorPipe] Always complete futures from thread pool
* #38931 [TensorPipe] Do not mark future messages as complete after they have timed out
* #38929 [TensorPipe] Fix lock inversion upon response read error handling
* #38928 [TensorPipe] Fix timeout computation
* #38927 [TensorPipe] Add cases for TP in RPC test helpers
* #38926 [TensorPipe] Pass names of endpoints to context/pipe for easier debugging

The `test_rref_context_debug_info` test was flaky with the TensorPipe agent, and I think the issue is the test itself.

What was happening is that on line 1826 the test was clearing a global variable on the remote side which was holding a rref. Even though the RPC call that unset the global variable was synchronous, the messages that the rref context needs to send around to delete that rref are asynchronous. Therefore, sometimes, when we reached line 1845 we saw the following check fail:
```
        self.assertEqual(2, int(info["num_owner_rrefs"]))
```
because `num_owner_rrefs` was still 3, as the deletion hadn't yet been processed.

The only way I found to fix it is to add a synchronization step where we wait for all the futures from the rref context to complete. Since we must wait for this to happen on all workers, we synchronize with a barrier.

Differential Revision: [D21716070](https://our.internmc.facebook.com/intern/diff/D21716070/)

lw committed:
[RPC] Fix flaky test by waiting for async rref calls

The `test_rref_context_debug_info` test was flaky with the TensorPipe agent, and I think the issue is the test itself.

What was happening is that on line 1826 the test was clearing a global variable on the remote side which was holding a rref. Even though the RPC call that unset the global variable was synchronous, the messages that the rref context needs to send around to delete that rref are asynchronous. Therefore, sometimes, when we reached line 1845 we saw the following check fail:
```
        self.assertEqual(2, int(info["num_owner_rrefs"]))
```
because `num_owner_rrefs` was still 3, as the deletion hadn't yet been processed.

The only way I found to fix it is to add a synchronization step where we wait for all the futures from the rref context to complete. Since we must wait for this to happen on all workers, we synchronize with a barrier.

Differential Revision: [D21716070](https://our.internmc.facebook.com/intern/diff/D21716070/)

[ghstack-poisoned]

lw committed:
Update on "[RPC] Fix flaky test by waiting for async rref calls"


The `test_rref_context_debug_info` test was flaky with the TensorPipe agent, and I think the issue is the test itself.

What was happening is that on line 1826 the test was clearing a global variable on the remote side which was holding a rref. Even though the RPC call that unset the global variable was synchronous, the messages that the rref context needs to send around to delete that rref are asynchronous. Therefore, sometimes, when we reached line 1845 we saw the following check fail:
```
        self.assertEqual(2, int(info["num_owner_rrefs"]))
```
because `num_owner_rrefs` was still 3, as the deletion hadn't yet been processed.

The only way I found to fix it is to add a synchronization step where we wait for all the futures from the rref context to complete. Since we must wait for this to happen on all workers, we synchronize with a barrier.

Differential Revision: [D21716070](https://our.internmc.facebook.com/intern/diff/D21716070/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/rref_context.cpp:
```diff
@@ -175,11 +177,15 @@ void RRefContext::delUser(
       // Sending an RRefUserDelete causes the receiver to run delForkOfOwner,
       // which is now idempotent. See the comment at RRefContext::delForkOfOwner
       // for more details.
+      ++numPendingFutures_;
```
This looks OK to me. I wonder would it be better if we make this a guard with a move ctor?

lw replied:
As you prefer. We could indeed create a small type, which increases this count in the constructor and decreases it in the destructor, and capture an object of that class in the closure.

mrshenli replied:
no need to change in this PR. And honestly there is no clear winner which one is better.

mrshenli commented on torch/csrc/distributed/rpc/rref_context.h:
```diff
@@ -285,6 +285,8 @@ class TORCH_API RRefContext {
   std::unordered_map<ForkId, c10::intrusive_ptr<RRef>, ForkId::Hash>
       pendingChildren_;
 
+  std::atomic<int64_t> numPendingFutures_{0};
```
Let's add some comments to explain this metrics

lw replied:
Done.

mrshenli commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -1811,9 +1811,9 @@ def test_rref_context_debug_info(self):
         rpc.rpc_sync(worker_name(dst_rank), set_global_rref, args=(rref1,))
 
         # barrier before check 2
+        wait_until_pending_futures_and_users_flushed()
```
Hmm, this can make tests pass, but what about real-world applications? Do we need to expose this API, maybe as a private one (start with `_`) and with a better name. :)

lw replied:
Do we expect users to care about this? It seems to me that it is part of the design that the rref operations are async and therefore "lagging" a bit behind. Do we see reasons for the users to want to wait for these and thus force a synchronization?

mrshenli replied:
No, I am not aware of any use case as of today.

mrshenli reviewed (commented):
Thanks for adding the metrics!

lw committed:
Update on "[RPC] Fix flaky test by waiting for async rref calls"


The `test_rref_context_debug_info` test was flaky with the TensorPipe agent, and I think the issue is the test itself.

What was happening is that on line 1826 the test was clearing a global variable on the remote side which was holding a rref. Even though the RPC call that unset the global variable was synchronous, the messages that the rref context needs to send around to delete that rref are asynchronous. Therefore, sometimes, when we reached line 1845 we saw the following check fail:
```
        self.assertEqual(2, int(info["num_owner_rrefs"]))
```
because `num_owner_rrefs` was still 3, as the deletion hadn't yet been processed.

The only way I found to fix it is to add a synchronization step where we wait for all the futures from the rref context to complete. Since we must wait for this to happen on all workers, we synchronize with a barrier.

Differential Revision: [D21716070](https://our.internmc.facebook.com/intern/diff/D21716070/)

[ghstack-poisoned]

lw committed:
Update on "[RPC] Fix flaky test by waiting for async rref calls"


The `test_rref_context_debug_info` test was flaky with the TensorPipe agent, and I think the issue is the test itself.

What was happening is that on line 1826 the test was clearing a global variable on the remote side which was holding a rref. Even though the RPC call that unset the global variable was synchronous, the messages that the rref context needs to send around to delete that rref are asynchronous. Therefore, sometimes, when we reached line 1845 we saw the following check fail:
```
        self.assertEqual(2, int(info["num_owner_rrefs"]))
```
because `num_owner_rrefs` was still 3, as the deletion hadn't yet been processed.

The only way I found to fix it is to add a synchronization step where we wait for all the futures from the rref context to complete. Since we must wait for this to happen on all workers, we synchronize with a barrier.

Differential Revision: [D21716070](https://our.internmc.facebook.com/intern/diff/D21716070/)

[ghstack-poisoned]
